usaspending_api/etl/elasticsearch_loader_helpers/transform_data.py

```diff
@@ -0,0 +1,49 @@
+import logging
+
+from django.conf import settings
+from time import perf_counter
+
+from usaspending_api.etl.elasticsearch_loader_helpers.utilities import (
+    convert_postgres_json_array_to_list,
+    format_log,
+)
+
+
+logger = logging.getLogger("script")
+
+
+def transform_award_data(worker, records):
+    converters = {}
+    return transform_data(worker, records, converters)
+
+
+def transform_transaction_data(worker, records):
+    converters = {
+        "federal_accounts": convert_postgres_json_array_to_list,
+    }
+    return transform_data(worker, records, converters)
+
+
+def transform_data(worker, records, converters):
+    logger.info(format_log(f"Transforming data", job=worker.name, process="Index"))
+    start = perf_counter()
+
+    for record in records:
+        for field, converter in converters.items():
```

Comment:
You are correct on the function overhead of not using pandas, but in this branch there is only a single "conversion" function and a few dictionary key additions which all together takes sub 1s for each partition, hard to justify spending a lot of time on this sub-second step when the SQL extraction takes 10+ minutes per partition. Seth's changes in his PR DO require more functions and currently that does take 40-50% of each partition ETL when using materialized data instead of the current SQL views. There could be A/B testing of Pandas vs native datatypes after his changes are included and the database data are staged for efficient extraction.

I wouldn't assume that in this specific usage that pandas would automatically be more memory efficient. The original code converted most of the data columns in the DataFrame to strings to ensure the data stored in Elasticsearch was correct. Storing a bunch of unique strings isn't efficient in pandas: https://pandas.pydata.org/pandas-docs/stable/user_guide/scale.html#use-efficient-datatypes


The data provided to the elasticsearch batch indexer are lists of dictionaries. I'm assuming they are converted to JSON when sent to Elasticsearch. Might be a small hit converting a pandas dataframe to a list of dictionaries but I wouldn't expect that to be too time intensive, might require extra memory while both data structures are present in memory